### PR TITLE
Added map for two_floors_house and a filter for the Virtual Wall messages

### PR DESCRIPTION
**Pull request by stevendes**
virtual_wall_steven

**UPDATE**

- Added a map for the first floor of the two_floors_house world

- Added a filter for the messages sent by the virtual walls, avoiding the overlap of messages

**ADDED VIRTUAL WALL TO THE STATE MACHINE**

- Modified the `smach_ranking_controller` to work with virtual walls

**HOW TO TEST IT**

- Launch the two_floors_map with the number of robots you want to spawn with the `NUM_ROBOTS` parameter Ex: `NUM_ROBOTS=4 roslaunch ca_gazebo create_two_floors_house.launch`

- Launch the smach controller with the namespace of the robot you want to control. Ex: `roslaunch ca_node smach_controller.launch ns:="create2"`

### CODE REVIEW

File: ca_gazebo/include/ca_gazebo/virtual_wall_detector.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        public:
+
+        ///Constructor
+        GazeboVirtualWallDetector();   
+
+        ///Destructor
+        ~GazeboVirtualWallDetector();
+
+    /// \brief Load the plugin
```
Comment (eborghi10) [inline ca_gazebo/include/ca_gazebo/virtual_wall_detector.h]: Check tabulations.

File: ca_gazebo/include/ca_gazebo/virtual_wall_detector.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    /// Pointer to the model
+        physics::ModelPtr model;
+
+    /// Pointer to the update event connection
+        private: event::ConnectionPtr updateConnection;
+
+    /// Initialize ROS variables
+        ros::NodeHandle nh_;
+        ros::Publisher virtual_wall_publisher_;
+        ros::Subscriber wall_subscriber_;
```
Comment (eborghi10) [inline ca_gazebo/include/ca_gazebo/virtual_wall_detector.h]: Using `wall` or `virtual_wall` everywhere is redundant.

File: ca_gazebo/include/ca_gazebo/virtual_wall_detector.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        ros::Time prev_update_time_;
+        event::ConnectionPtr updateConnection_;
+        
+    ///  Auxiliar variables
+        bool virtual_wall_status;
+        bool virtual_wall_sub;
```
Comment (eborghi10) [inline ca_gazebo/include/ca_gazebo/virtual_wall_detector.h]: Please choose a better name. Think about what this variable does.

File: ca_gazebo/launch/create_two_floors_house.launch [hunk 4]
```diff
@@ -1,7 +1,7 @@
 <launch>
   <arg name="virtual_wall_x" default="0.2"/>
   <arg name="virtual_wall_y" default="-1.2"/>
-  <arg name="virtual_wall_yaw" default="1.57"/>
+  <arg name="virtual_wall_yaw" default="-1.57"/>
```
Comment (eborghi10) [inline ca_gazebo/launch/create_two_floors_house.launch:4]: Was this change necessary at the end? I don't think so. Please double check.

File: ca_gazebo/src/virtual_wall_detector.cpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  {
+    // Make sure the ROS node for Gazebo has already been initialized
+    if (!ros::isInitialized())
+    {
+      ROS_FATAL_STREAM("A ROS node for Gazebo has not been initialized, unable to load plugin. "
+        << "Load the Gazebo system plugin 'libgazebo_ros_api_plugin.so' in the gazebo_ros package)");
+      return;
+    }
+    // Initialize the ros variables and gaebo variables
+    this->virtual_wall_publisher_ = nh_.advertise<std_msgs::Bool>("/filtered/virtual_wall", 1);
```
Comment (eborghi10) [inline ca_gazebo/src/virtual_wall_detector.cpp]: The topic name should be `/createN/virtual_wall` and the original topic should be `/createN/virtual_wall/raw`.

File: ca_gazebo/src/virtual_wall_detector.cpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    this->prev_update_time_ = ros::Time::now();
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////
+  // Update the controller
+  void GazeboVirtualWallDetector::OnUpdate()
+  //Every update check if one of the sensors is active, if it is, change the output to true, publishes every 1ms
+  {
+    this->virtual_wall_status += this->virtual_wall_sub;
+    if ((ros::Time::now() - this->prev_update_time_) < ros::Duration(0.1))
```
Comment (eborghi10) [inline ca_gazebo/src/virtual_wall_detector.cpp]: `ros::Duration(0.1)` should be a SDF parameter than can be modifiable.

File: ca_gazebo/src/virtual_wall_detector.cpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    this->virtual_wall_publisher_.publish(this->virtual_wall_status);
+    this->virtual_wall_status = false;
+    this->prev_update_time_ = ros::Time::now();
+  }
+
+  void GazeboVirtualWallDetector::WallCallback(const std_msgs::Bool::ConstPtr& data)
+  // Callback to get the data from the virtual wall sensors
+  {
+    this->virtual_wall_sub = data->data;
```
Comment (eborghi10) [inline ca_gazebo/src/virtual_wall_detector.cpp]: Recommendation: Use a mutex here.

**Review by eborghi10 — CHANGES_REQUESTED** [review]
First pass done, PTAL.

File: ca_gazebo/launch/create_two_floors_house.launch (hunk 4 above)
Comment (stevendes) [reply ca_gazebo/launch/create_two_floors_house.launch:4]: In the context of what the virtual_wall is supposed to do it's necessary to block the correct opening of the room. Not doing this makes having a virtual wall blocking a real wall and not an opening.

**Comment by stevendes** [comment]
Addressed all comments and ready for review!

File: ca_gazebo/src/create_virtual_wall_plugin.cpp [hunk 8]
```diff
@@ -119,7 +119,7 @@ void VirtualWallSensorPlugin::OnUpdate()
     const ignition::math::Pose3d & robot_pose = model->WorldPose();
     sensor_offset = robot_pose.Rot().RotateVector(sensor_offset);
     ignition::math::Pose3d sensor_pose = robot_pose + sensor_pose;
-    ROS_INFO_NAMED("virtual_wall_plugin", "sensor_pose x=%f y=%f z=%f",
+    ROS_DEBUG_NAMED("virtual_wall_plugin", "sensor_pose x=%f y=%f z=%f",
```
Comment (eborghi10) [inline ca_gazebo/src/create_virtual_wall_plugin.cpp:122]: Was this change on purpose?

File: ca_gazebo/src/create_virtual_wall_plugin.cpp [hunk 9]
```diff
@@ -130,12 +130,12 @@ void VirtualWallSensorPlugin::OnUpdate()
     double phi = std::acos(error.Dot(laser_dir) / error.Length());
     double d = error.Length() * std::sin(phi);
     double lambda = (sensor_pose.Pos().X() - laser_pose.Pos().X()) / laser_dir.X();
-    ROS_INFO_NAMED("virtual_wall_plugin", "d=%f lambda=%f detected_range=%f error=%f",
+    ROS_DEBUG_NAMED("virtual_wall_plugin", "d=%f lambda=%f detected_range=%f error=%f",
```
Comment (eborghi10) [inline ca_gazebo/src/create_virtual_wall_plugin.cpp:133]: Was this change on purpose?

File: ca_gazebo/src/create_virtual_wall_plugin.cpp [hunk 10]
```diff
@@ -130,12 +130,12 @@ void VirtualWallSensorPlugin::OnUpdate()
...
         d,
         lambda,
         detected_range,
         error.Length());
-
+
```
Comment (eborghi10) [inline ca_gazebo/src/create_virtual_wall_plugin.cpp]: Revert this change.

File: ca_gazebo/src/virtual_wall_detector.cpp [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+      int argc = 0;
+      char** argv = NULL;
+      ros::init(argc, argv, "virtual_wall_detector");
+    }
+    // Initialize the ros variables and gaebo variables
+    const std::string & name = _parent->GetName();
+    std::istringstream iss(name.substr(create2_model_name_prefix_length));
+    size_t x;
+    iss >> x;
+//    ROS_WARN("The name %s",name.c_str());
```
Comment (eborghi10) [inline ca_gazebo/src/virtual_wall_detector.cpp]: Remove.

File: ca_gazebo/src/virtual_wall_detector.cpp [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    if (iss && iss.eof()) 
+    {
+      std::ostringstream oss;
+      oss << "create" << x << "/virtual_wall/raw/";
+      std::string topic_name_ = oss.str();
+      ROS_WARN("The name %s",oss.str().c_str());
```
Comment (eborghi10) [inline ca_gazebo/src/virtual_wall_detector.cpp]: Is this necessary?

File: ca_gazebo/src/virtual_wall_detector.cpp [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    const std::string & name = _parent->GetName();
+    std::istringstream iss(name.substr(create2_model_name_prefix_length));
+    size_t x;
+    iss >> x;
+//    ROS_WARN("The name %s",name.c_str());
+    if (iss && iss.eof()) 
+    {
+      std::ostringstream oss;
+      oss << "create" << x << "/virtual_wall/raw/";
+      std::string topic_name_ = oss.str();
```
Comment (eborghi10) [inline ca_gazebo/src/virtual_wall_detector.cpp]: Make it `const` in case you don't have plans to modify it.

File: ca_gazebo/src/virtual_wall_detector.cpp [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+      ROS_WARN("The name %s",oss.str().c_str());
+      this->subscriber_ = this->nh_.subscribe(topic_name_, 1, &GazeboVirtualWallDetector::WallCallback, this);
+      oss.str("");
+      oss.clear();
```
Comment (eborghi10) [inline ca_gazebo/src/virtual_wall_detector.cpp]: Is this necessary?

File: ca_gazebo/src/virtual_wall_detector.cpp (hunk 13 above)
Comment (eborghi10) [inline ca_gazebo/src/virtual_wall_detector.cpp]: Local variables must not have an ending `_`. **Be consistent with the coding style.**

File: ca_gazebo/src/virtual_wall_detector.cpp [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+      oss << "create" << x << "/virtual_wall/";
+      topic_name_ = oss.str();
+      this->publisher_ = this->nh_.advertise<std_msgs::Bool>(topic_name_, 1);
+      this->updateConnection_ = event::Events::ConnectWorldUpdateBegin(
+        std::bind(&GazeboVirtualWallDetector::OnUpdate, this));    
+      this->prev_update_time_ = ros::Time::now();
+      this->update_period_ = 1.0/(_sdf->Get<double>("updateRate"));
```
Comment (eborghi10) [inline ca_gazebo/src/virtual_wall_detector.cpp]: This is a bad to put this code. Do you know why?

Move it as the first step in `Load()`.

**Review by eborghi10 — CHANGES_REQUESTED** [review]
I made more comments, PTAL.

File: ca_gazebo/src/create_virtual_wall_plugin.cpp (hunk 8 above)
Comment (stevendes) [reply ca_gazebo/src/create_virtual_wall_plugin.cpp:122]: Yes, that is the information of the virtual walls that by default were printing in the console every update, so you can't really see the gazebo startup information because they are quickly displaced by a barrage of messages showing the values of every virtual wall.

File: ca_gazebo/src/create_virtual_wall_plugin.cpp (hunk 9 above)
Comment (stevendes) [reply ca_gazebo/src/create_virtual_wall_plugin.cpp:133]: Same as above.

File: ca_gazebo/src/virtual_wall_detector.cpp (hunk 13 above)
Comment (stevendes) [reply ca_gazebo/src/virtual_wall_detector.cpp]: I'm using the same variable here to get the name of both topics, subscriber and publisher.

File: ca_gazebo/src/virtual_wall_detector.cpp (hunk 15 above)
Comment (stevendes) [reply ca_gazebo/src/virtual_wall_detector.cpp]: I don't know why exactly.

**Comment by stevendes** [comment]
@eborghi10 ready for testing

File: ca_gazebo/include/ca_gazebo/virtual_wall_detector.h [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    /// \brief Callback to the virtual wall sensors
+    /// \param take pointer to the message
+        void WallCallback(const std_msgs::Bool::ConstPtr& data); 
+
+    /// \brief Update the controller
+        void OnUpdate();
+
+        private:
+
+    /// Pointer to the model
```
Comment (eborghi10) [inline ca_gazebo/include/ca_gazebo/virtual_wall_detector.h]: Why are you using different tabulation for comments and code?

File: ca_gazebo/src/virtual_wall_detector.cpp [hunk 17]
```diff
@@ -0,0 +1,72 @@
+#include "ca_gazebo/virtual_wall_detector.h"
+
+#include <gazebo/common/Plugin.hh>
```
Comment (eborghi10) [inline ca_gazebo/src/virtual_wall_detector.cpp]: move this to the header file.

**Review by eborghi10 — CHANGES_REQUESTED** [review]
I couldn't find the issue with the compilation. Let's talk about this tomorrow.

**Comment by eborghi10** [comment]
I fixed the compilation issues that I was having. I don't know why you don't see them:

```
diff --git a/ca_gazebo/src/virtual_wall_detector.cpp b/ca_gazebo/src/virtual_wall_detector.cpp
index 250dac1..52e1fcc 100644
--- a/ca_gazebo/src/virtual_wall_detector.cpp
+++ b/ca_gazebo/src/virtual_wall_detector.cpp
@@ -57,7 +57,9 @@ namespace gazebo
     {
       return;
     }
-    this->publisher_.publish(this->is_vwall_detected_);
+    std_msgs::Bool msg;
+    msg.data = this->is_vwall_detected_;
+    this->publisher_.publish(msg);
     this->is_vwall_detected_ = false;
     this->prev_update_time_ = ros::Time::now();
   }
```

File: ca_gazebo/src/virtual_wall_detector.cpp [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    }
+    this->publisher_.publish(this->is_vwall_detected_);
+    this->is_vwall_detected_ = false;
+    this->prev_update_time_ = ros::Time::now();
+  }
+
+  void GazeboVirtualWallDetector::WallCallback(const std_msgs::Bool::ConstPtr& data)
+  // Callback to get the data from the virtual wall sensors
+  {
+    mtx.lock();
```
Comment (eborghi10) [inline ca_gazebo/src/virtual_wall_detector.cpp]: This is not the way that we use mutexes, the idea is to use "scoped mutexes". Let's talk about this during the meeting.

**Review by eborghi10 — CHANGES_REQUESTED** [review]

File: ca_gazebo/src/virtual_wall_detector.cpp (hunk 18 above)
Comment (eborghi10) [reply ca_gazebo/src/virtual_wall_detector.cpp]: [https://en.cppreference.com/w/cpp/thread/lock_guard](https://en.cppreference.com/w/cpp/thread/lock_guard)

**Comment by eborghi10** [comment]
One thing that I detected in the simulation:
- The robot can move the virtual wall from its position so I suggest you to make that model static.

The robot is detecting the virtual wall in this condition and that's bad. **The virtual wall is there to protect the robot from entering the other room.**
image

The virtual wall is also being detected here and that's okay.
image

But the correct position of the virtual wall should be this (where the device is pointing to the door).
image

Steps that I did to debug (only one robot and moved it with the keyboard):

```
roslaunch ca_gazebo create_two_floors_house.launch

rostopic echo /create1/virtual_wall

roslaunch ca_tools keyboard_teleop.launch id:=1
```

File: ca_gazebo/include/ca_gazebo/virtual_wall_detector.h [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+        ros::Publisher publisher_;
+        ros::Subscriber subscriber_;
+        ros::Time prev_update_time_;
+        event::ConnectionPtr updateConnection_;
+        
+        /// Auxiliar variables
+        bool is_vwall_detected_;
+        bool get_vwall_;
+        double update_period_;
+        std::mutex vwall_mutex;
```
Comment (eborghi10) [inline ca_gazebo/include/ca_gazebo/virtual_wall_detector.h:52]: The naming convention for this and other variables don't match with the rest. Keep the same around all the code.

File: ca_node/scripts/smach_ranking_controller.py [hunk 20]
```diff
@@ -110,7 +120,8 @@ def main():
                                transitions={'left_pressed':'ROTATE_RIGHT_SEQ', 
                                             'right_pressed':'ROTATE_LEFT_SEQ',
                                             # 'both_pressed': 'BACKWARD'})
-                                            'both_pressed': 'ROTATE_LEFT_SEQ'})
+                                            'both_pressed': 'ROTATE_LEFT_SEQ',
+                                            'virtual_wall_detected': 'ROTATE_LEFT_SEQ'})
```
Comment (eborghi10) [inline ca_node/scripts/smach_ranking_controller.py:124]: What is the expected behavior when the robot detects a virtual wall message?

**Review by eborghi10 — CHANGES_REQUESTED** [review]
